crates/bevy_app/src/app.rs

```diff
@@ -357,6 +359,62 @@ impl App {
         self
     }
 
+    /// Adds a [`State<S>`] and [`NextState<S>`] whenever entering a matching `Parent` state
+    /// from a `Parent` state that does not match and runs OnEnter/OnExit when those occur.
+    ///
+    /// In addition, adds an instance of [`apply_state_transition::<S>`] in
+    /// [`StateTransition`] so that transitions happen before [`Update`](crate::Update).
+    ///
+    /// You can treat these just like any other states, but bear in mind that the `State<S>` and `NextState<S>`
+    /// resources might not exist.
+    pub fn add_sub_state<S: States, Parent: States>(
+        &mut self,
+        condition: impl StateMatcher<Parent>,
+    ) -> &mut Self {
+        self.add_systems(
+            Parent::on_enter_matching(condition.clone()),
+            initialize_state_and_enter::<S>,
+        )
+        .add_systems(
+            Parent::on_exit_matching(condition.clone()),
+            remove_state_from_world::<S>,
+        )
+        .add_systems(
+            StateTransition,
+            apply_state_transition::<S>.run_if(in_state(condition)),
+        );
+
+        self
+    }
+
+    /// Adds a [`State<S>`] and [`NextState<S>`] whenever entering a matching `Parent` state
+    /// regardless of whether the previous `Parent` state matches, and runs OnEnter/OnExit when those occur.
```

Comment:
This appears to be the critical distinction here. It's a bit hard to follow from the docs, but more importantly, it's not clear why we would want to have the option of both behaviors.

Reply:
If I'm understanding correctly, the issue is the distinction of strict vs. not strict in general (i.e. not just in sub states), and why we would want both? I will continue my response with that assumption, but please let me know if I'm mistaken.

I'll give a few examples here:

- Turn sub-state that might occur in different higher-level situations, but needs to be reset every time we switch
```rust
enum Turn {
  #[default]
  DrawCards,
  PlayCards,
  Attack,
  EnemyDefense,
  Resolve
}

enum GameState {
  PlayerTurn,
  EnemyTurn,
  MatchEnded
}

...
// this isn't current syntax, just an illustration
app.add_strict_sub_state::<Turn, _>(state_matches!(GameState, PlayerTurn | EnemyTurn))
```

Here, if I had a non-strict sub state, the Turn won't be re-set to default when moving between `PlayerTurn` and `EnemyTurn`.

- A "Pause" state, on the other hand, should be re-set every time the GameState changes, but should still only exist when pausing is sensible (i.e. we're actively in game)

```rust
enum Pause {
  #[default]
  Running,
  Paused
}

...

app.add_sub_state::<Pause, _>(state_matches!(GameState, PlayerTurn | EnemyTurn))
```

In non-sub-state contexts, this can also be relevant. Say I have a function I want to clean up the UI whenever I move between screens - I can set it up with:
```rust
enum AppState {
  MainMenu,
  OptionMenu,
  InGame
}

.add_system(on_exit_strict!(AppState, MainMenu | OptionMenu), ui_cleanup_system))
```
because it's strict, it will run every time I transition out of a matching state, even if the incoming state is also a match.

However, I might have a backdrop animated thing in those menus that I want to remain consistent, so I would use the non-strict version for that:

```rust
.add_system(on_enter!(AppState, MainMenu | OptionMenu), setup_menu_backdrop)
.add_system(on_exit!(AppState, MainMenu | OptionMenu), remove_menu_backdrop)
```

Do these examples clarify the need here? Should I create some more specific examples for these situations rather than the current setup of the state example?